File: src/services/EmbeddingService.ts
export interface WordEmbedding {
  word: string;
  embedding: number[];
  dimension: number;
}

export interface SentenceEmbedding {
  text: string;
  embedding: number[];
  dimension: number;
}

export interface SimilarWord {
  word: string;
  similarity: number;
}

export interface SemanticSearchResult {
  index: number;
  text: string;
  similarity: number;
  text_preview: string;
}

export interface EmbeddingVisualization {
  plot: string; // JSON string of Plotly figure
  method: string;
  words_count: number;
  words_found: string[];
  words_not_found: string[];
}

export interface EmbeddingModel {
  id: string;
  type: string;
  path: string;
  config: any;
  vocabulary_size: number;
  trained_on: number;
  created_at: string;
}

export interface EmbeddingStats {
  vocabulary_size: number;
  vector_size: number;
  most_frequent_words: [string, number][];
  model_type: string;
}

export interface SemanticAnalysis {
  text: string;
  sentence_embedding_shape: number[];
  sentence_embedding_norm: number;
  word_count: number;
  unique_words: number;
  words_in_vocabulary: number;
  word_similarities: { [key: string]: [string, number][] };
  semantic_density: number;
}

import ConfigManager from '../config/AppConfig';

export class EmbeddingService {
  private static readonly BASE_URL = ConfigManager.getApiUrl('embeddings');

  // Entraîner un modèle TF-IDF
  static async trainTFIDF(texts: string[]): Promise<{ stats: any }> {
    try {
      const response = await fetch(`${this.BASE_URL}/train/tfidf`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ texts }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de l\'entraînement TF-IDF');
      }

      return { stats: data.stats };
    } catch (error) {
      console.error('Erreur entraînement TF-IDF:', error);
      throw error;
    }
  }

  // Entraîner un modèle Word2Vec (utilise TF-IDF pour l'instant)
  static async trainWord2Vec(texts: string[], config: any = {}): Promise<EmbeddingModel> {
    try {
      // Utiliser l'endpoint TF-IDF existant comme alternative
      const response = await fetch(`${this.BASE_URL}/train/tfidf`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ texts }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de l\'entraînement');
      }

      // Retourner un modèle formaté
      return {
        id: 'tfidf_' + Date.now(),
        type: 'tfidf',
        path: '',
        config: config,
        vocabulary_size: data.stats?.vocabulary_size || 0,
        trained_on: texts.length,
        created_at: new Date().toISOString()
      };
    } catch (error) {
      console.error('Erreur entraînement Word2Vec:', error);
      throw error;
    }
  }

  // Obtenir l'embedding d'un mot
  static async getWordEmbedding(word: string, modelId?: string): Promise<WordEmbedding> {
    try {
      const params = new URLSearchParams();
      if (modelId) params.append('model_id', modelId);

      const response = await fetch(`${this.BASE_URL}/word/${encodeURIComponent(word)}?${params}`);

      if (!response.ok) {
        if (response.status === 404) {
          throw new Error(`Mot "${word}" non trouvé dans le vocabulaire`);
        }
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la récupération de l\'embedding');
      }

      return {
        word: data.word,
        embedding: data.embedding,
        dimension: data.dimension
      };
    } catch (error) {
      console.error('Erreur récupération embedding mot:', error);
      throw error;
    }
  }

  // Obtenir l'embedding d'une phrase
  static async getSentenceEmbedding(text: string): Promise<SentenceEmbedding> {
    try {
      const response = await fetch(`${this.BASE_URL}/sentence`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ text }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la récupération de l\'embedding');
      }

      return {
        text: data.text,
        embedding: data.embedding,
        dimension: data.dimension
      };
    } catch (error) {
      console.error('Erreur récupération embedding phrase:', error);
      throw error;
    }
  }

  // Trouver des mots similaires
  static async findSimilarWords(word: string, topK: number = 10, modelId?: string): Promise<SimilarWord[]> {
    try {
      const params = new URLSearchParams();
      params.append('top_k', topK.toString());
      if (modelId) params.append('model_id', modelId);

      const response = await fetch(`${this.BASE_URL}/similar/${encodeURIComponent(word)}?${params}`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la recherche de similarité');
      }

      return data.similar_words.map(([word, similarity]: [string, number]) => ({
        word,
        similarity
      }));
    } catch (error) {
      console.error('Erreur recherche mots similaires:', error);
      throw error;
    }
  }

  // Recherche sémantique
  static async semanticSearch(query: string, texts: string[], topK: number = 5): Promise<SemanticSearchResult[]> {
    try {
      const response = await fetch(`${this.BASE_URL}/search`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ query, texts, top_k: topK }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la recherche sémantique');
      }

      return data.results;
    } catch (error) {
      console.error('Erreur recherche sémantique:', error);
      throw error;
    }
  }

  // Visualiser les embeddings
  static async visualizeEmbeddings(
    texts: string[],
    labels?: string[],
    method: string = 'pca'
  ): Promise<EmbeddingVisualization> {
    const response = await fetch(`${this.BASE_URL}/visualize`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        texts,
        labels,
        method
      }),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ error: 'Unknown error' }));
      throw new Error(errorData.error || `HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data.visualization;
  }

  // Analyser la sémantique d'un texte
  static async analyzeTextSemantics(text: string): Promise<SemanticAnalysis> {
    try {
      const response = await fetch(`${this.BASE_URL}/analyze`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ text }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de l\'analyse sémantique');
      }

      return data.analysis;
    } catch (error) {
      console.error('Erreur analyse sémantique:', error);
      throw error;
    }
  }

  // Obtenir la liste des modèles
  static async getEmbeddingModels(): Promise<EmbeddingModel[]> {
    try {
      const response = await fetch(`${this.BASE_URL}/models`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la récupération des modèles');
      }

      return data.models;
    } catch (error) {
      console.error('Erreur récupération modèles:', error);
      throw error;
    }
  }

  // Obtenir les statistiques d'un modèle
  static async getEmbeddingStats(modelId?: string): Promise<EmbeddingStats> {
    try {
      const response = await fetch(`${this.BASE_URL}/status`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      if (!data.success) {
        throw new Error(data.error || 'Erreur lors de la récupération des statistiques');
      }

      // Retourner des statistiques par défaut basées sur le statut
      return {
        vocabulary_size: data.vocabulary_size || 0,
        vector_size: data.vector_size || 0,
        most_frequent_words: data.most_frequent_words || [],
        model_type: data.model_type || 'tfidf'
      };
    } catch (error) {
      console.error('Erreur récupération statistiques:', error);
      // Retourner des statistiques par défaut en cas d'erreur
      return {
        vocabulary_size: 0,
        vector_size: 0,
        most_frequent_words: [],
        model_type: 'tfidf'
      };
    }
  }

  // Vérifier si le service d'embedding est disponible
  static async isEmbeddingServiceAvailable(): Promise<boolean> {
    try {
      const response = await fetch(`${this.BASE_URL}/models`);
      return response.ok;
    } catch (error) {
      return false;
    }
  }
} 

File: backend/models/embedding_service.py
import numpy as np
import pandas as pd
from gensim.models import Word2Vec
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
import umap
import plotly.graph_objects as go
import plotly.express as px
import json
import os
from datetime import datetime
from typing import List, Dict, Tuple, Optional, Any
import pickle
import re

class EmbeddingService:
    def __init__(self):
        self.word2vec_model = None
        self.sentence_transformer = None
        self.embeddings_cache = {}
        self.models_dir = "./models/embeddings"
        os.makedirs(self.models_dir, exist_ok=True)
        
        # Charger les modèles pré-entraînés
        self._load_pretrained_models()
    
    def _load_pretrained_models(self):
        """Charge les modèles pré-entraînés"""
        try:
            # Modèle Sentence-BERT pour les embeddings de phrases
            print("Chargement du modèle Sentence-BERT...")
            self.sentence_transformer = SentenceTransformer('all-MiniLM-L6-v2')
            print("✅ Sentence-BERT chargé avec succès")
        except Exception as e:
            print(f"⚠️ Erreur chargement Sentence-BERT: {e}")
    
    def train_word2vec(self, texts: List[str], config: Dict = None) -> Dict:
        """Entraîne un modèle Word2Vec sur les textes fournis"""
        try:
            if config is None:
                config = {}
            
            # Préprocessing des textes
            processed_texts = []
            for text in texts:
                # Nettoyage et tokenisation
                text = re.sub(r'[^\w\s]', '', text.lower())
                tokens = text.split()
                if len(tokens) > 0:
                    processed_texts.append(tokens)
            
            if len(processed_texts) == 0:
                raise ValueError("Aucun texte valide pour l'entraînement")
            
            # Configuration par défaut
            default_config = {
                'vector_size': 100,
                'window': 5,
                'min_count': 2,
                'workers': 4,
                'epochs': 10,
                'sg': 1  # Skip-gram
            }
            default_config.update(config)
            
            # Entraînement Word2Vec
            print(f"Entraînement Word2Vec sur {len(processed_texts)} textes...")
            self.word2vec_model = Word2Vec(
                sentences=processed_texts,
                vector_size=default_config['vector_size'],
                window=default_config['window'],
                min_count=default_config['min_count'],
                workers=default_config['workers'],
                epochs=default_config['epochs'],
                sg=default_config['sg']
            )
            
            # Sauvegarder le modèle
            model_id = datetime.now().strftime("%Y%m%d_%H%M%S")
            model_path = f"{self.models_dir}/word2vec_{model_id}.model"
            self.word2vec_model.save(model_path)
            
            # Métadonnées
            model_info = {
                'id': model_id,
                'type': 'word2vec',
                'path': model_path,
                'config': default_config,
                'vocabulary_size': len(self.word2vec_model.wv.key_to_index),
                'trained_on': len(processed_texts),
                'created_at': datetime.now().isoformat()
            }
            
            # Sauvegarder les métadonnées
            with open(f"{self.models_dir}/word2vec_{model_id}_info.json", 'w') as f:
                json.dump(model_info, f, indent=2)
            
            return model_info
            
        except Exception as e:
            raise Exception(f"Erreur entraînement Word2Vec: {str(e)}")
    
    def get_word_embedding(self, word: str, model_id: str = None) -> Optional[np.ndarray]:
        """Obtient l'embedding d'un mot"""
        try:
            if model_id:
                # Charger un modèle spécifique
                model_path = f"{self.models_dir}/word2vec_{model_id}.model"
                if os.path.exists(model_path):
                    model = Word2Vec.load(model_path)
                    if word in model.wv:
                        return model.wv[word]
            else:
                # Utiliser le modèle courant
                if self.word2vec_model and word in self.word2vec_model.wv:
                    return self.word2vec_model.wv[word]
            
            return None
        except Exception as e:
            print(f"Erreur récupération embedding: {e}")
            return None
    
    def get_sentence_embedding(self, text: str) -> np.ndarray:
        """Obtient l'embedding d'une phrase avec Sentence-BERT"""
        try:
            if self.sentence_transformer is None:
                raise Exception("Sentence-BERT non disponible")
            
            # Cache pour éviter les recalculs
            if text in self.embeddings_cache:
                return self.embeddings_cache[text]
            
            embedding = self.sentence_transformer.encode(text)
            self.embeddings_cache[text] = embedding
            
            return embedding
        except Exception as e:
            raise Exception(f"Erreur embedding phrase: {str(e)}")
    
    def find_similar_words(self, word: str, top_k: int = 10, model_id: str = None) -> List[Tuple[str, float]]:
        """Trouve les mots les plus similaires"""
        try:
            model = self.word2vec_model
            
            if model_id:
                model_path = f"{self.models_dir}/word2vec_{model_id}.model"
                if os.path.exists(model_path):
                    model = Word2Vec.load(model_path)
            
            if model is None:
                return []
            
            if word not in model.wv:
                return []
            
            similar_words = model.wv.most_similar(word, topn=top_k)
            return similar_words
            
        except Exception as e:
            print(f"Erreur recherche similarité: {e}")
            return []
    
    def semantic_search(self, query: str, texts: List[str], top_k: int = 5) -> List[Dict]:
        """Recherche sémantique dans une collection de textes"""
        try:
            if self.sentence_transformer is None:
                raise Exception("Sentence-BERT non disponible")
            
            # Embedding de la requête
            query_embedding = self.get_sentence_embedding(query)
            
            # Embeddings des textes
            text_embeddings = []
            for text in texts:
                embedding = self.get_sentence_embedding(text)
                text_embeddings.append(embedding)
            
            # Calcul des similarités
            similarities = cosine_similarity([query_embedding], text_embeddings)[0]
            
            # Tri par similarité décroissante
            results = []
            for i, (text, similarity) in enumerate(zip(texts, similarities)):
                results.append({
                    'index': i,
                    'text': text,
                    'similarity': float(similarity),
                    'text_preview': text[:200] + "..." if len(text) > 200 else text
                })
            
            # Retourner les top_k résultats
            results.sort(key=lambda x: x['similarity'], reverse=True)
            return results[:top_k]
            
        except Exception as e:
            raise Exception(f"Erreur recherche sémantique: {str(e)}")
    
    def visualize_embeddings(self, words: List[str], method: str = 'tsne', model_id: str = None) -> Dict:
        """Visualise les embeddings de mots en 2D"""
        try:
            model = self.word2vec_model
            
            if model_id:
                model_path = f"{self.models_dir}/word2vec_{model_id}.model"
                if os.path.exists(model_path):
                    model = Word2Vec.load(model_path)
            
            if model is None:
                raise Exception("Aucun modèle Word2Vec disponible")
            
            # Récupérer les embeddings des mots disponibles
            embeddings = []
            valid_words = []
            
            for word in words:
                if word in model.wv:
                    embeddings.append(model.wv[word])
                    valid_words.append(word)
            
            if len(embeddings) == 0:
                raise Exception("Aucun mot trouvé dans le vocabulaire")
            
            embeddings = np.array(embeddings)
            
            # Réduction de dimensionnalité
            if method == 'pca':
                reducer = PCA(n_components=2, random_state=42)
            elif method == 'tsne':
                reducer = TSNE(n_components=2, random_state=42, perplexity=min(30, len(embeddings)-1))
            elif method == 'umap':
                reducer = umap.UMAP(n_components=2, random_state=42)
            else:
                raise Exception(f"Méthode non supportée: {method}")
            
            embeddings_2d = reducer.fit_transform(embeddings)
            
            # Créer le graphique Plotly
            fig = go.Figure()
            
            fig.add_trace(go.Scatter(
                x=embeddings_2d[:, 0],
                y=embeddings_2d[:, 1],
                mode='markers+text',
                text=valid_words,
                textposition='top center',
                marker=dict(
                    size=10,
                    color=np.arange(len(valid_words)),
                    colorscale='Viridis',
                    showscale=True
                ),
                hovertemplate='<b>%{text}</b><br>X: %{x:.2f}<br>Y: %{y:.2f}<extra></extra>'
            ))
            
            fig.update_layout(
                title=f'Visualisation des Embeddings ({method.upper()})',
                xaxis_title=f'{method.upper()} Dimension 1',
                yaxis_title=f'{method.upper()} Dimension 2',
                showlegend=False,
                template='plotly_dark'
            )
            
            return {
                'plot': fig.to_json(),
                'method': method,
                'words_count': len(valid_words),
                'words_found': valid_words,
                'words_not_found': [w for w in words if w not in valid_words]
            }
            
        except Exception as e:
            raise Exception(f"Erreur visualisation: {str(e)}")
    
    def get_embedding_statistics(self, model_id: str = None) -> Dict:
        """Obtient les statistiques d'un modèle d'embedding"""
        try:
            model = self.word2vec_model
            
            if model_id:
                model_path = f"{self.models_dir}/word2vec_{model_id}.model"
                if os.path.exists(model_path):
                    model = Word2Vec.load(model_path)
            
            if model is None:
                return {'error': 'Aucun modèle disponible'}
            
            # Statistiques du vocabulaire
            vocab_size = len(model.wv.key_to_index)
            vector_size = model.wv.vector_size
            
            # Mots les plus fréquents
            word_counts = [(word, model.wv.get_vecattr(word, "count")) 
                          for word in list(model.wv.key_to_index.keys())[:20]]
            word_counts.sort(key=lambda x: x[1], reverse=True)
            
            return {
                'vocabulary_size': vocab_size,
                'vector_size': vector_size,
                'most_frequent_words': word_counts[:10],
                'model_type': 'Word2Vec'
            }
            
        except Exception as e:
            return {'error': str(e)}
    
    def get_available_models(self) -> List[Dict]:
        """Retourne la liste des modèles d'embedding disponibles"""
        models = []
        
        try:
            for filename in os.listdir(self.models_dir):
                if filename.endswith('_info.json'):
                    info_path = os.path.join(self.models_dir, filename)
                    with open(info_path, 'r') as f:
                        model_info = json.load(f)
                        models.append(model_info)
        except Exception as e:
            print(f"Erreur chargement modèles: {e}")
        
        return models
    
    def analyze_text_semantics(self, text: str) -> Dict:
        """Analyse sémantique complète d'un texte"""
        try:
            # Embedding de la phrase
            sentence_embedding = self.get_sentence_embedding(text)
            
            # Analyse des mots individuels
            words = re.findall(r'\b\w+\b', text.lower())
            word_embeddings = {}
            word_similarities = {}
            
            if self.word2vec_model:
                for word in set(words):
                    if word in self.word2vec_model.wv:
                        word_embeddings[word] = self.word2vec_model.wv[word].tolist()
                        # Similarités avec d'autres mots du texte
                        similarities = []
                        for other_word in set(words):
                            if other_word != word and other_word in self.word2vec_model.wv:
                                sim = self.word2vec_model.wv.similarity(word, other_word)
                                similarities.append((other_word, float(sim)))
                        word_similarities[word] = sorted(similarities, key=lambda x: x[1], reverse=True)[:3]
            
            return {
                'text': text,
                'sentence_embedding_shape': sentence_embedding.shape,
                'sentence_embedding_norm': float(np.linalg.norm(sentence_embedding)),
                'word_count': len(words),
                'unique_words': len(set(words)),
                'words_in_vocabulary': len(word_embeddings),
                'word_similarities': word_similarities,
                'semantic_density': len(word_embeddings) / len(set(words)) if len(set(words)) > 0 else 0
            }
            
        except Exception as e:
            raise Exception(f"Erreur analyse sémantique: {str(e)}") 